### ISSUE TITLE
[Seine/Ganges][pdx201/kirin] VoLTE still refuse to work.

### ISSUE THREAD

**Issue by AngryGami** [sonyxperiadev/bug_tracker#659]
Platform: Seine and Ganges
Device: pdx201 and kirin
Kernel version: 4.14
Android version: 10
Software binaries version: latest available for both devices

Previously working on
Maybe on Android 9, but I'm not sure.

**Description**
After resolving #640 I still can't make VoLTE work on either of these devices. Now I can see that VoLTE feature is available on device, enabled and provisioned (whatever that means) but when I go into service menu `*#*#4636#*#*` -> `three dots on top right side of the view` -> `IMS service status` - It still says:
```
IMS Registration: Not registered
Voice over LTE: Unavailable
Voice over WiFi: Unavailable
Video Calling: Unavailable
UT Interface: Unavailable
```

Logs now contain IMS service failed attempts to register:
```
2020-11-20 12:18:37.088 1985-1985/com.android.phone I/QImsService: ImsSenderRxr : [0032]> REQUEST_GET_IMS_CONFIG [SUB0]
2020-11-20 12:18:37.089 1985-1985/com.android.phone I/QImsService: ImsSenderRxr : REQUEST_GET_IMS_CONFIG request to ImsRadio: token 32 item:25 boolValue:false intValue:0 strValue:null errorCause:0
2020-11-20 12:18:37.095 1985-2121/com.android.phone I/QImsService: ImsRadioResponse : Get config response received.
2020-11-20 12:18:37.095 1985-2121/com.android.phone I/QImsService: IFRequest : [0032]< REQUEST_GET_IMS_CONFIG error: 2
2020-11-20 12:18:37.096 1985-2411/com.android.phone I/QImsService: ImsConfigImplHandler : Message received: what = 1
2020-11-20 12:18:37.097 1985-2411/com.android.phone I/QImsService: ImsConfigImpl : Notifyall
```

I must note that same sim-card used in my personal Samsung 8 device does allow VoLTE calls and I've tried different cell providers as well with same results - i.e. does work with my Samsung and doesn't with Xperia 10 Mark 1/2 SODP build.

Could that be problem with Sony binaries? Or there is some issue with my cell provider?

**How to reproduce**
Build image with re-signed Sony odm applications as in #640, Put SIM-card to a device try to make VoLTE call.

**Comment by sjllls** [comment]
I met this issue again, but it seems VoLTE works fine on LineageOS 17.1.

**Comment by AngryGami** [comment]
@sjllls Can you please check if `vendor.qti.iwlan` package is actually installed? (`adb shell pm list packages vendor.qti.iwlan`)

**Comment by AngryGami** [comment]
Well no matter what I do I can't make `vendor.qti.iwlan` package get installed. Here are messages related to IWlanService:
```
2020-11-20 23:14:47.133 832-1249/? I/ServiceManagement: Registered vendor.qti.hardware.data.iwlan@1.0::IIWlan/slot1 (start delay of 2482ms)
2020-11-20 23:14:48.019 834-1349/? I/ServiceManagement: Registered vendor.qti.hardware.data.iwlan@1.0::IIWlan/slot2 (start delay of 3358ms)
2020-11-20 23:14:49.744 1400-1400/system_process I/PackageManager: /odm/app/IWlanService changed; collecting certs
2020-11-20 23:14:49.751 1400-1400/system_process W/PackageManager: Failed to scan /odm/app/IWlanService: Reconcile failed: Signature mismatch for shared user: SharedUserSetting{d82ca05 android.uid.phone/1001}
2020-11-20 23:14:49.840 1400-1400/system_process I/PackageManager: /odm/app/IWlanService changed; collecting certs
2020-11-20 23:14:49.843 1400-1400/system_process W/PackageManager: Failed to scan /product/app/IWlanService: Reconcile failed: Signature mismatch for shared user: SharedUserSetting{d82ca05 android.uid.phone/1001}
2020-11-20 23:14:53.381 1400-1768/system_process W/ActivityManager: Unable to start service Intent { act=android.telephony.data.QualifiedNetworksService pkg=vendor.qti.iwlan } U=0: not found
2020-11-20 23:14:53.388 1400-2011/system_process W/ActivityManager: Unable to start service Intent { act=android.telephony.NetworkService pkg=vendor.qti.iwlan } U=0: not found
2020-11-20 23:14:53.461 1400-2011/system_process E/DefaultPermGrantPolicy: PackageNot found: vendor.qti.iwlan
    android.content.pm.PackageManager$NameNotFoundException: vendor.qti.iwlan
        at android.app.ApplicationPackageManager.getPackageInfoAsUser(ApplicationPackageManager.java:190)
        at android.app.ApplicationPackageManager.getPackageInfo(ApplicationPackageManager.java:159)
        at com.android.server.pm.permission.DefaultPermissionGrantPolicy.getPackageInfo(DefaultPermissionGrantPolicy.java:1317)
        at com.android.server.pm.permission.DefaultPermissionGrantPolicy.getPackageInfo(DefaultPermissionGrantPolicy.java:1308)
        at com.android.server.pm.permission.DefaultPermissionGrantPolicy.isSystemPackage(DefaultPermissionGrantPolicy.java:963)
        at com.android.server.pm.permission.DefaultPermissionGrantPolicy.grantPermissionsToSystemPackage(DefaultPermissionGrantPolicy.java:364)
        at com.android.server.pm.permission.DefaultPermissionGrantPolicy.grantSystemFixedPermissionsToSystemPackage(DefaultPermissionGrantPolicy.java:350)
        at com.android.server.pm.permission.DefaultPermissionGrantPolicy.grantDefaultPermissionsToEnabledTelephonyDataServices(DefaultPermissionGrantPolicy.java:848)
        at com.android.server.pm.PackageManagerService.lambda$grantDefaultPermissionsToEnabledTelephonyDataServices$36$PackageManagerService(PackageManagerService.java:25256)
        at com.android.server.pm.-$$Lambda$PackageManagerService$Jk56TooKBJVbCXzf9mkL2I94Ey0.runOrThrow(Unknown Source:6)
        at android.os.Binder.withCleanCallingIdentity(Binder.java:357)
        at com.android.server.pm.PackageManagerService.grantDefaultPermissionsToEnabledTelephonyDataServices(PackageManagerService.java:25255)
        at android.content.pm.IPackageManager$Stub.onTransact(IPackageManager.java:4346)
        at com.android.server.pm.PackageManagerService.onTransact(PackageManagerService.java:4059)
        at android.os.Binder.execTransactInternal(Binder.java:1021)
        at android.os.Binder.execTransact(Binder.java:994)
```
Though package is there:
```
pdx201:/product/app/IWlanService # ls -la
total 276
drwxr-xr-x  2 root shell   4096 2020-09-10 16:16 .
drwxr-xr-x 22 root root    4096 2020-11-21 00:12 ..
-rw-r--r--  1 root root  270743 2020-09-10 16:16 IWlanService.apk
```
And it is signed with `product` key:
```
include $(CLEAR_VARS)
LOCAL_MODULE := IWlanService
LOCAL_SRC_FILES := IWlanService.apk
LOCAL_CERTIFICATE := platform
LOCAL_MODULE_TAGS := optional
LOCAL_PRODUCT_MODULE := true
LOCAL_MODULE_CLASS := APPS
LOCAL_MODULE_SUFFIX := $(COMMON_ANDROID_PACKAGE_SUFFIX)
include $(BUILD_PREBUILT)
```
I've double-cheked this:
```
pdx201:/data/system # grep -A 3 'android.uid.phone' packages.xml
    <shared-user name="android.uid.phone" userId="1001">
        <sigs count="1" schemeVersion="3">
            <cert index="2" />
        </sigs>
```
```
pdx201-hp:/data/system # grep -B 3 -F '<cert index="2"' packages.xml | grep -A 4 'ims'
    <package name="org.codeaurora.ims" codePath="/product/priv-app/ims" nativeLibraryPath="/product/priv-app/ims/lib" primaryCpuAbi="arm64-v8a" publicFlags="537443909" privateFlags="524296" ft="175e62801d8" it="175e62801d8" ut="175e62801d8" version="1" sharedUserId="1001" isOrphaned="true">
        <sigs count="1" schemeVersion="3">
            <cert index="2" />
```
shared user references certificate with index 2, the same index package `org.codeaurora.ims` references and this package is signed with same key as IWlanService.apk:
```
include $(CLEAR_VARS)
LOCAL_MODULE := ims
LOCAL_SRC_FILES := ims.apk
LOCAL_CERTIFICATE := platform
LOCAL_MODULE_TAGS := optional
LOCAL_PRODUCT_MODULE := true
LOCAL_MODULE_CLASS := APPS
LOCAL_MODULE_SUFFIX := $(COMMON_ANDROID_PACKAGE_SUFFIX)
LOCAL_PRIVILEGED_MODULE := true
include $(BUILD_PREBUILT)
```

What is going on here?

**Comment by MarijnS95** [comment]
```
2020-11-20 23:14:49.751 1400-1400/system_process W/PackageManager: Failed to scan /odm/app/IWlanService: Reconcile failed: Signature mismatch for shared user: SharedUserSetting{d82ca05 android.uid.phone/1001}
```
Perhaps this is conflicting?

**Comment by MarijnS95** [comment]
> I've double-cheked this:

Note also that you're looking at `<package name="org.codeaurora.ims" codePath="/product/priv-app/ims"` where the error clearly surrounds `/product/app/IWlanService` instead.

**Comment by sjllls** [comment]
@AngryGami 
```
[sjll@sjll desktop]$ adb shell pm list packages vendor.qti.iwlan
package:vendor.qti.iwlan
```

I tried to make a clean build, VolTE works now.

If you included https://github.com/PixelExperience-Devices/device_sony_customization

I think you will meet https://github.com/Havoc-OS/android_build/commit/4a93012f351edeaa1140875157f5d1697ade3378#diff-c3e87a7526d039953580e4c6d2d26fbca9426e39d7bf4be73b78a212a77bd6e3

Log:
```
11-21 09:31:24.465  1349  1349 W PackageManager: Skipping provider name androidx.lifecycle.process.lifecycle-process (in package com.android.systemui): name already used by com.android.settings
11-21 09:31:24.483  1349  1349 W PackageManager: Failed to scan /product/app/IWlanService: Application package vendor.qti.iwlan already installed.  Skipping duplicate.
11-21 09:31:24.558  1349  1689 W PackageParser: No actions in intent filter at /product/app/CalendarGooglePrebuilt/CalendarGooglePrebuilt.apk Binary XML file line #500
11-21 09:31:24.588  1349  1349 W PackageManager: Failed to scan /product/app/CneApp: Application package com.qualcomm.qti.cne already installed.  Skipping duplicate.
```

**Comment by AngryGami** [comment]
@MarijnS95 
> Perhaps this is conflicting?

This is in odm partition, I can't remove that. Other packages that do get installed also produce this kind of messages.

> Note also that you're looking at <package name="org.codeaurora.ims" codePath="/product/priv-app/ims" where the error clearly surrounds /product/app/IWlanService instead.

I've given `org.codeaurora.ims` as an example of a package that has same signature, also lives in odm partition, was overriden by version from `product` and was successfully installed. I.e. `ims.apk` get installed no problem while `IWlanService.apk` does not.

@sjllls 
Thanks. Apparently on your device this package get installed and I would like to know how and why. Is your build uses custom keys for signing?
I wonder where your version of `vendor.qti.iwlan` package is installed from could you please show what `adb shell pm list packages -a -f vendor.qti.iwlan` prints out for your system?

I can't "include" device_sony_customizations, I had to adopt whatever it tries to do into my build. I took `Android.mk` file from that repo and bunch of apk files from my device and put them under `vendor/sonyodm` directory in my AOSP sources tree. Then I've included applications that defined in `Android.mk` into my custom version of `aosp_xqau52.mk`, added copy `emptyfile` command and permissions for ims package. I don't think I need anything else from these customizations (e.g. Hexagon app, libs for soundtrigger, wallpaper or widevine), though I'm not 100% sure. For example do I need `TARGET_USES_AOSP_APNS_CONF` or `DEXPREOPT_GENERATE_APEX_IMAGE ` or `TARGET_KERNEL_HEADERS`?

**Comment by sjllls** [comment]
> @MarijnS95
> 
> > Perhaps this is conflicting?
> 
> This is in odm partition, I can't remove that. Other packages that do get installed also produce this kind of messages.
> 
> > Note also that you're looking at <package name="org.codeaurora.ims" codePath="/product/priv-app/ims" where the error clearly surrounds /product/app/IWlanService instead.
> 
> I've given `org.codeaurora.ims` as an example of a package that has same signature, also lives in odm partition, was overriden by version from `product` and was successfully installed. I.e. `ims.apk` get installed no problem while `IWlanService.apk` does not.
> 
> @sjllls
> Thanks. Apparently on your device this package get installed and I would like to know how and why. Is your build uses custom keys for signing?
> I wonder where your version of `vendor.qti.iwlan` package is installed from could you please show what `adb shell pm list packages -a -f vendor.qti.iwlan` prints out for your system?
> 
> I can't "include" device_sony_customizations, I had to adopt whatever it tries to do into my build. I took `Android.mk` file from that repo and bunch of apk files from my device and put them under `vendor/sonyodm` directory in my AOSP sources tree. Then I've included applications that defined in `Android.mk` into my custom version of `aosp_xqau52.mk`, added copy `emptyfile` command and permissions for ims package. I don't think I need anything else from these customizations (e.g. Hexagon app, libs for soundtrigger, wallpaper or widevine), though I'm not 100% sure. For example do I need `TARGET_USES_AOSP_APNS_CONF` or `DEXPREOPT_GENERATE_APEX_IMAGE ` or `TARGET_KERNEL_HEADERS`?

No problem, I will feedback to you tomorrow.

**Comment by sjllls** [comment]
@AngryGami  I built SODP AOSP, so I can include customization directly. You might need to do much more works.
```
[sjll@sjll desktop]$ adb shell pm list packages -a -f vendor.qti.iwlan
package:/product/app/IWlanService/IWlanService.apk=vendor.qti.iwlan
```

**Comment by MarijnS95** [comment]
> > Perhaps this is conflicting?
> 
> This is in odm partition, I can't remove that. Other packages that do get installed also produce this kind of messages.

That's odd, both `CneApp` and `IWlanService` are in `/odm/app` as well as `/odm/system_ext/app`. Conveniently exactly those two placeholder folders were missing result in https://github.com/sonyxperiadev/device-sony-common/pull/801 (however the bindmounts were there).

Given that this results in the same error I believe it doesn't really matter where it's used from. We should remove one or the other (coordinated with the removal from odm).

Are you sure you signed and replaced the correct apk? Both apks in `/odm/app` and `/product/app` give the same ` Signature mismatch for shared user` error,

> I've given `org.codeaurora.ims` as an example of a package that has same signature, also lives in odm partition, was overriden by version from `product` and was successfully installed. I.e. `ims.apk` get installed no problem while `IWlanService.apk` does not.

Since `IWlanService.apk` is causing trouble while `ims.apk` is fine it'd be insightful to see the matching xml for the wlan app, in particular what signature is selected.

> For example do I need `TARGET_USES_AOSP_APNS_CONF` or `DEXPREOPT_GENERATE_APEX_IMAGE ` or `TARGET_KERNEL_HEADERS`?

I don't think you need any of those, but you need `TARGET_USES_ODM_APPS_BINDMOUNT := false` to turn off the bindmounting script which is completely superseded by the PE script.

**Comment by AngryGami** [comment]
@MarijnS95 
I've pulled `IWlanService.apk` from my device and noticed that it has some strange signature (`CERT.RSA` file was strange while digests were fine). I've rebuild it and verified that new image contains apk with correct file and now it actually get installed during initial boot. Though this doesn't seem to have any impact on VoLTE - it still doesn't work and `IMS Registration` is still unregistered. I've noticed new exception in the logs though:
```
11-20 22:50:47.326  3011  4179 W QImsService: ImsConfigImpl : Ignore VoWiFi Roaming enable/disable
11-20 22:50:47.326  3011  4179 D ImsConfigImplBase: Set provision value of 26 to 0 failed with error code 1
11-20 22:50:47.326  3011  4179 D ImsConfig: setConfig(): item = 26 value = 0 ret = 1
11-20 22:50:47.329  3011  3011 D SubscriptionController: Valid row not present in db
11-20 22:50:47.329  3011  3011 D SubscriptionController: getSubscriptionProperty Query value = null
11-20 22:50:47.330  1224  1299 I commit_sys_config_file: [notification-policy,11]
11-20 22:50:47.330  3011  4177 I QImsService: ImsSenderRxr : [0022]> REQUEST_SET_IMS_CONFIG [SUB1]
11-20 22:50:47.330  3011  4177 I QImsService: ImsSenderRxr : REQUEST_SET_IMS_CONFIG request to ImsRadio: token 22 item:29 boolValue:false intValue:1 strValue:null errorCause:0
11-20 22:50:47.331  3011  3011 D ImsManager: updateVideoCallFeatureValue: available = true, enabled = true, nonTTY = true, data enabled = false
11-20 22:50:47.332  3011  3011 I ImsManager: update RTT: settings value: false always-on carrierconfig: false
11-20 22:50:47.332  3011  3011 D ImsManager: updateUtFeatureValue: available = false, isProvisioned = true, enabled = false
11-20 22:50:47.333  3011  3011 I ImsManager: changeMmTelCapability: changing capabilities for sub: -1, request: CapabilityChangeRequest{mCapabilitiesToEnable={CapabilityPair{mCapability=1, radioTech=0}, CapabilityPair{mCapability=2, radioTech=0}}, mCapabilitiesToDisable={CapabilityPair{mCapability=1, radioTech=1}, CapabilityPair{mCapability=4, radioTech=0}}}
11-20 22:50:47.333  3011  3669 I QImsService: ImsRadioResponse : Set config response received.
11-20 22:50:47.333  3011  3669 I QImsService: IFRequest : [0022]< REQUEST_SET_IMS_CONFIG error: 2
11-20 22:50:47.333  3011  3011 E ImsManager: updateImsServiceConfig: 
11-20 22:50:47.333  3011  3011 E ImsManager: com.android.ims.ImsException: changeMmTelCapability(CCR)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at com.android.ims.ImsManager.changeMmTelCapability(ImsManager.java:2144)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at com.android.ims.ImsManager.updateImsServiceConfig(ImsManager.java:1437)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at com.android.internal.telephony.GsmCdmaPhone.handleMessage(GsmCdmaPhone.java:2503)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at android.os.Handler.dispatchMessage(Handler.java:107)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at android.os.Looper.loop(Looper.java:214)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at android.app.ActivityThread.main(ActivityThread.java:7356)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at java.lang.reflect.Method.invoke(Native Method)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
11-20 22:50:47.333  3011  3011 E ImsManager: Caused by: android.os.RemoteException: ImsServiceProxy is not ready to accept commands.
11-20 22:50:47.333  3011  3011 E ImsManager: 	at com.android.ims.MmTelFeatureConnection.checkServiceIsReady(MmTelFeatureConnection.java:962)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at com.android.ims.MmTelFeatureConnection.changeEnabledCapabilities(MmTelFeatureConnection.java:753)
11-20 22:50:47.333  3011  3011 E ImsManager: 	at com.android.ims.ImsManager.changeMmTelCapability(ImsManager.java:2128)
11-20 22:50:47.333  3011  3011 E ImsManager: 	... 8 more
11-20 22:50:47.333  3011  3887 I QImsService: ImsConfigImplHandler : Message received: what = 1
11-20 22:50:47.333  3011  3887 I QImsService: ImsConfigImpl : Notifyall
11-20 22:50:47.333  3011  4177 D ImsConfigImplBase: Set provision value of 27 to 1 failed with error code 1
```

@sjllls 
Thanks again. I've resolved my issue with iwlan. You are saying that VoLTE works for you but can you tell what does service menu (one that I've described in first message) shows to you? Does it says "Voice over LTE: Available"?

**Comment by sjllls** [comment]
> 
> @sjllls
> Thanks again. I've resolved my issue with iwlan. You are saying that VoLTE works for you but can you tell what does service menu (one that I've described in first message) shows to you? Does it says "Voice over LTE: Available"?

@AngryGami  Yup, only Video Calling is no (The same on stock).

**Comment by AngryGami** [comment]
@sjllls  Thanks.
Crap.. I have no idea then why it doesn't work for me...

**Comment by AngryGami** [comment]
So I've flashed stock image onto my pdx201 device and VoLTE doesn't work:
![1](https://user-images.githubusercontent.com/2470910/99990926-b5a84880-2db4-11eb-92b7-28feb1baad6f.png)
![2](https://user-images.githubusercontent.com/2470910/99990929-b6d97580-2db4-11eb-961d-232d33255b13.png)
![3](https://user-images.githubusercontent.com/2470910/99990933-b7720c00-2db4-11eb-8329-e8bc4c6104aa.png)

@sjllls can you please compare this with what you see on your device? I kinda stuck and don't understand what is happening with all this stuff.

**Comment by AngryGami** [comment]
And now I can't boot my phone with SODP image (after I reflashed it from stock) it manages to boot to adb available but then just stuck with:
```
01-01 05:47:19.880   722   722 E PDC_UTIL: qmi_client_init returned failure (-3) for PDC, retry times 2
01-01 05:47:24.888   722   722 E PDC_UTIL: qmi_client_init returned failure (-3) for PDC, retry times 3
01-01 05:47:27.278     0     0 E i2c_geni 4a88000.i2c: i2c error :-107
01-01 05:47:29.897   722   722 E PDC_UTIL: qmi_client_init returned failure (-3) for PDC, retry times 4
01-01 05:47:34.905   722   722 E PDC_UTIL: qmi_client_init returned failure (-3) for PDC, retry times 5
01-01 05:47:39.913   722   722 E PDC_UTIL: qmi_client_init returned failure (-3) for PDC, retry times 6
```

**Comment by AngryGami** [comment]
I've also noticed following behavior. After initial flashing of device `vendor.qti.iwlan` package get installed (though VoLTE still doesn't work), but after reboot it get uninstalled with following messages in the log:
```
11-23 21:42:59.536  1393  1393 I PackageManager: /odm/app/IWlanService changed; collecting certs
11-23 21:42:59.542  1393  1393 I PackageManager: Update system package vendor.qti.iwlan code path from /product/app/IWlanService to /odm/app/IWlanService; Retain data and using new
11-23 21:42:59.542  1393  1393 I PackageManager: Update system package vendor.qti.iwlan resource path from /product/app/IWlanService to /odm/app/IWlanService; Retain data and using new
11-23 21:42:59.543  1393  1393 W PackageManager: Failed to scan /odm/app/IWlanService: Reconcile failed: Signature mismatch for shared user: SharedUserSetting{b516440 android.uid.phone/1001}
...
11-23 21:43:00.124  1393  1393 I PackageManager: /odm/app/IWlanService changed; collecting certs
11-23 21:43:00.128  1393  1393 I PackageManager: Update system package vendor.qti.iwlan code path from /product/app/IWlanService to /odm/app/IWlanService; Retain data and using new
11-23 21:43:00.128  1393  1393 I PackageManager: Update system package vendor.qti.iwlan resource path from /product/app/IWlanService to /odm/app/IWlanService; Retain data and using new
11-23 21:43:00.130  1393  1393 W PackageManager: Failed to scan /product/app/IWlanService: Reconcile failed: Signature mismatch for shared user: SharedUserSetting{b516440 android.uid.phone/1001}
...
11-23 21:43:00.250  1393  1393 W PackageManager: System package vendor.qti.iwlan no longer exists; it's data will be wiped
11-23 21:43:00.250  1393  1393 I pm_critical_info: System package vendor.qti.iwlan no longer exists; it's data will be wiped
```
Same happens with `com.qualcomm.qti.cne` package.

**Comment by MarijnS95** [comment]
https://github.com/sonyxperiadev/bug_tracker/issues/659#issuecomment-731759654

> We should remove one or the other (coordinated with the removal from odm).

**Comment by sjllls** [comment]
@AngryGami  Did you flash stock firmware with EMMA? And are you in UK?

**Comment by AngryGami** [comment]
@sjllls Yes I flashed stock with EMMA. I'm not in the UK, though stock image is for UK as far as I can tell:
![image](https://user-images.githubusercontent.com/2470910/100068510-f1382680-2e37-11eb-8ba3-ccbaefc3a086.png)
These are only options I have for flashing and I can't even flash `A.6.24` - it always give me error in the middle of flashing process.

**Comment by AngryGami** [comment]
@MarijnS95 I don't understand what are you pointing to.

**Comment by MarijnS95** [comment]
@AngryGami Remove IWlanService and cne from `/odm/app`.

**Comment by AngryGami** [comment]
Ah.. so no solution for me.

**Comment by sjllls** [comment]
@AngryGami  This might be the reason, Sony stored its IMS configs in OEM partition, different areas have specific configs.
I'm not sure if pdx201 could enable VoLTE without stock data in modem partition on SODP. But you might need to flash correct FW to enable VoLTE on stock at first.

You could try newflasher, but it doesn't works for me at all. I had to relock my bootloader and repair with xperia companion. But you flashed with emma, now it might be treated as a UK model. I'm not sure it will restore your phone to your area's FW....

**Comment by sjllls** [comment]
I also have IWlanService in product and odm partition at the same time. But there is no error about signature.

**Comment by MarijnS95** [comment]
> Ah.. so no solution for me.

@AngryGami You can always alter odm locally and see if that solves anything. My guess based on the logs is that the apps are "fighting" each other: one is installed but AOSP tries to use the other one with a different signature, causing havoc (as confirmed by the logs again).

@sjllls Curious, are the apps in `/odm` even being discovered? Is there anything about it in the log _at all_?

**Comment by AngryGami** [comment]
@sjllls 
> I also have `IWlanService` in `product` and `odm` partition at the same time. But there is no error about signature.

This I can't explain... unless signatures of files on `odm` and `product` are the same somehow... though how it didn't work for you before you added PE customizations? Maybe as MarjinS95 suggest - it wasn't even trying to install from `odm` somehow.

> You could download FW for your area with XperiaFirm, them unpack oem.sin, flash the ext4 image via
fastboot.

I'll try that.

**Comment by AngryGami** [comment]
@MarijnS95 Ok, I've manage to alter sony image and deleted `/odm/app` from there. Now apps are not conflicting and stays installed after reboot. Though VoLTE still doesn't work :) 
```
11-24 15:14:09.890  2036  2036 D ImsPhoneCallTracker: [0] handleFeatureCapabilityChanged: isVolteEnabled=false, isVideoCallEnabled=false, isVowifiEnabled=false, isUtEnabled=false
```

**Comment by MarijnS95** [comment]
@AngryGami Great, that's at least one more variable ruled out.

Did you manage to get it working on stock in the end? That's the first indicator things are awry. After that comparing logs with a device that has it working is probably easiest. I know right about nothing about IMS and VoLTE so can't help much here, sorry.

**Comment by AngryGami** [comment]
@MarijnS95 
Kirin device image available is Android 9 only (at least via Emma) and there VoLTE works. So no Android 10 stock for it and no definitive answer regarding VoLTE. Comparing logs from Android 9 and Android 10 is hard (I've tried) especially when I don't know where to look.
pdx201 stock image that I've tried (UK one, the only one Emma tool allow me to flash) is Android 10 and VoLTE doesn't work there. Not only it doesn't work - it also is not available as "enable/disable" option in mobile network settings menu. I assume this is just blocked there and I probably had to try to flash some other stock image. I've downloaded one using XperiaFirm, though I don't know how to flash it yet (if you have any advice about that I would like to hear it :)).

**Comment by sjllls** [comment]
> @MarijnS95
> Kirin device image available is Android 9 only (at least via Emma) and there VoLTE works. So no Android 10 stock for it and no definitive answer regarding VoLTE. Comparing logs from Android 9 and Android 10 is hard (I've tried) especially when I don't know where to look.
> pdx201 stock image that I've tried (UK one, the only one Emma tool allow me to flash) is Android 10 and VoLTE doesn't work there. Not only it doesn't work - it also is not available as "enable/disable" option in mobile network settings menu. I assume this is just blocked there and I probably had to try to flash some other stock image. I've downloaded one using XperiaFirm, though I don't know how to flash it yet (if you have any advice about that I would like to hear it :)).

@AngryGami I think you only need to flash OEM image, other images is not necessary I think.

**Comment by AngryGami** [comment]
> I think you only need to flash OEM image, other images are not necessary.

@sjllls I don't understand what you mean by that. I was talking about devices flashed with stock images and I do not control what partitions will be flashed by Emma tool.
So far I haven't found any way to flash stock image that I've downloaded with XperiaFirm tool. Flashtool from here http://www.flashtool.net doesn't seem to recognize them.

**Comment by AngryGami** [comment]
I've tried several different stock images for pdx201:
```
(flashed using newflasher tool)
XQ-AU52_Customized EU_1321-6453_59.0.A.8.45_R24A 
XQ-AU52_EE UK_1321-8030_59.0.A.8.45_R14A
XQ-AU52_Vodafone LC Not Net Locked_59.0.A.8.45-R10A
(flashed with Emma)
FSP_PDX201_Customized-UK_GENERIC_59.0.A.1.296_1321-9381_R4A
```
And on none of these VoLTE works or even available as an option in settings menu. IMS service status is always "Not Registered".

@sjllls could you please tell what exactly you did to make it work? Did you use SODP build or maybe Lineage OS? If you saw this working on stock which customization?

**Comment by sjllls** [comment]
> @sjllls I don't understand what you mean by that. I was talking about devices flashed with stock images and I >do not control what partitions will be flashed by Emma tool.
In the FW dir of XperiaFirm, there is a file named `oem_X-FLASH-CUST-XXXX.sin`, you only need flash this file with newflasher(If it works). Or unpack it to ext4 image, and flash it via fastboot.

As I said above, for IMS on stock you only need correct oem image for your area and carrier. No matter your firmware is UK/US/EU/CN/HK/TW..... If you flashed the correct oem.sin,IMS will works for you.

> 
@sjllls could you please tell what exactly you did to make it work? Did you use SODP build or maybe Lineage OS? If you saw this working on stock which customization?

For me, I am using China Mobile, I only need to flash oem.sin in HK FW, VoLTE will works for me.
**You need to download the suitable FW for your carrier.**

From my test result, If I want to use VoLTE on SODP based ROMs, I need to enable it on stock at first, then flash SODP ROMs.

If you want to build SODP ROMs, must include these IMS blobs.  
https://github.com/PixelExperience-Devices/device_sony_customization

**Comment by sjllls** [comment]
If you still can't understand, for example:
For XZ2, it sales and has a special version for China mainland market, but this firmware doesn't include GMS. If I want to use GMS, I could flash UK/US/EU firmware, but VoLTE will never works for me because these FW don't include my carrier config.
So I could flash these FW, then flash the oem.sin of CN FW, I will get a firmware has both VoLTE and GMS.

**Comment by AngryGami** [comment]
>As I said above, for IMS on stock you only need correct oem image for your area and carrier. No matter your firmware is UK/US/EU/CN/HK/TW..... If you flashed the correct oem.sin,IMS will works for you.

I've flashed entire stock FW which, I assume, include oem.sin files, I understand that this was an overkill but I don't think this matter. I've flashed UK/EU/Global versions but since I'm in Russia and none of these worked - none of these are correct for my region. Though there are no firmwares that look more suitable (for XQ-AU52 in XperiaFirm) - which should I select then?

> You need to download the suitable FW for your carrier.

How to find one - that is the question :) 

> I need to enable it on stock at first

Do you mean by "enable" that you had to flash oem.sin suitable for you region/carrier? Or something else?

> If you want to build SODP ROMs, must include these IMS blobs, or IMS will never work.

I did that already, not exactly by including device_sony_customization but by reproducing what it does in my build. So now my image includes all apk files from that repo and they did get installed properly.

**Comment by AngryGami** [comment]
> I need to enable it on stock at first, then flash SODP ROMs.

I've tried following:
1. I've flashed stock image (all partitions) "XQ-AU52_Customized EU_1321-6453_59.0.A.8.45_R24A" to a pdx201 device that actually in Norway (remotely)
2. Booted it and verified that it has VoLTE (which it does... or at least now IMS service is initialized and Voice over LTE: Accessible).
3. Then I flashed my SODP image (with my custom certificates)

After that phone is just stuck in "Sony" logo during boot.
I then flashed Sony SODP oem image from [here](https://developer.sony.com/develop/open-devices/downloads/software-binaries). Well not exactly it - one that I had to "fix" by removing apps from `/odm/apps` that were causing conflicts with iwlan and cne packages. Only after that phone actually can boot again, but no VoLTE for me. Which probably means image from Sony site doesn't work in Norway for some reason while stock image for EU does.

I assume that oem.sin file that you suggesting I should flash is flashed to same partition where Sony image does and therefore I can't make it work with my custom SODP.

**Comment by sjllls** [comment]
>since I'm in Russia and none of these worked

Hmm, this is the reason. RU has a special version on Xperia 1/5 II, but 10 II doesn't sales in RU I think. So there is no suitable oem.sin for your carrier.

But I still have method to enable it. Modify modem partition  manually.

Flash Customized RU 53.1.A.3.57 in your kirin(I4113), enable VoLTE on it.
Get root permission, and `adb pull /vendor/firmware_mnt`
Now search for RU configs in `firmware_mnt`.

For example, China country code is `CN` and my carrier is `China mobile` which code name is `CMCC`. So the config is 
`\vendor\firmware_mnt\image\modem_pr\mcfg\configs\mcfg_sw\generic\row\**cmcc**\vl\**cn**\mcfg_sw.mbn`
For another two China mainland carriers is `China Union` and `China Telecom`, so the code name is `cu` and `ct`
```
\vendor\firmware_mnt\image\modem_pr\mcfg\configs\mcfg_sw\generic\row\**cu**\vl\**cn**\mcfg_sw.mbn
\vendor\firmware_mnt\image\modem_pr\mcfg\configs\mcfg_sw\generic\row\**ct**\vl\**cn**\mcfg_sw.mbn
```
Flash your pdx201 to stock firmware and root it.
Copy these mbn files to your pdx201's modem partition,` vendor\firmware_mnt`
And add these mbn files in `\vendor\firmware_mnt\image\modem_pr\mcfg\configs\mcfg_sw\mbn_sw.txt`

Reboot, I think you will get IMS works.

**Comment by sjllls** [comment]
Okk, I know the reason now.
There is a overlay named `oem_overlay_com.sonymobile.customizationselector-res-305.apk` in oem partition of these phones before Xperia 1. Unpack it there is a `service_providers.xml`.
In there file:
```
    <service_provider_sim_config sim_config_id="S286.2">
        <mcc>655</mcc>
        <mnc>10</mnc>
    </service_provider_sim_config>
    <service_provider_sim_config sim_config_id="S287.1">
        <mcc>655</mcc>
        <mnc>07</mnc>
        <sp>Cell C</sp>
    </service_provider_sim_config>
```

It tells the customizationselector to choose the right mbn files in modem partition.
SODP also uses these mbn files to enable VoLTE.

All in all, if you don't have the right mbn files in modem partition, VoLTE will never work. But you are the unfortunate one. :(

So this issue could be closed, because there might be no solution except you try to hack these mbn files...

So this

**Comment by AngryGami** [comment]
> So this issue could be closed, because there might be no solution except you try to hack these mbn files...

Do you mean for Russia? I would like to fix that at least for Norway (with my custom SODP) - is there a way to do that?

**Comment by AngryGami** [comment]
So I've managed to find solution for Norway (and other Baltic countries I want).
Comment from @sjllls gave me last piece of the puzzle I needed. He mentioned these `mbn` files and I remembered where I've seen them before. Well not them exactly but these references like `mcfg_sw/generic/row/mts/vlvtvw/ru/mcfg_sw.mbn`. 
In the image there is a folder `/oem/modem-config` that contains bunch of sub-directories that look like this: `S11.3` `S136.1` `S140.3` etc. Each such directory contains `modem.conf` file and content of this file is exactly like this reference, e.g. `S11.3/modem.conf` contains `mcfg_sw/generic/euro/orange/vlvw/fr/mcfg_sw.mbn`.
I've took stock image for pdx201 device that I knew has VoLTE working (Customized EU one) and compared these directories with what SODP has. Stock image has files that mention `nor`, `telia`, `telenor` and `ice` operators while SODP doesn't, so I've took these files and added them to my SODP build. Here is full list:
```
S122.1
S125.1
S125.2
S249.1
S255.1
S260.1
S261.1
S54.1
S57.1
S58.1
S8604.1
S9093.1
S9161.1
S9162.2
S9162.3
```
These include also files for Sweden and Denmark.
I knew that `mnd` files referenced in these must be on modem partition because stock firmware works. When I've flashed my SODP images and "fixed" Sony binaries image (without duplicates of IWlan and Cne apps) - finally VoLTE worked.

I don't know though what @sjllls talks about with regards this `oem_overlay_com.sonymobile.customizationselector-res-305.apk` since I've not found nothing even closely lookalike on stock image for pdx201 (I've unpacked `sin` files and search there as well). But this doesn't matter :) since my method worked and I'm happy now.

I assume that I maybe even can enable VoLTE for Russia with similar approach, though this implies that modem partition actually contains files I want to reference, though this is not a priority.

**Comment by MarijnS95** [comment]
@AngryGami Thanks for the research!

That means we at least have your MCC/MNC listed in https://github.com/MarijnS95/SonyOpenTelephony/blob/master/ModemConfig/res/xml/service_provider_sim_configs.xml, otherwise it's another rabbit hole to go down.

Please leave this issue open and ask @jerpelea to include them in the next `odm` release, thanks!

**Comment by AngryGami** [comment]
Reopening, so maybe later this will be in `odm`.

**Comment by MarijnS95** [comment]
@AngryGami Sorry, I don't have my day today, please submit the missing files to https://github.com/sonyxperiadev/device-sony-common/tree/master/rootdir/vendor/oem/modem-config

**Comment by sjllls** [comment]
@AngryGami I mean if you are in Russia, VoLTE will never work on pdx201, because Sony didn't provide the mbn for this country. Unless you can make it yourself.

If you are in Norway, maybe there is suitable mbn in it. Good luck!

**Comment by sjllls** [comment]
@AngryGami  Does IMS work for you on pdx201 on AOSP 11?

**Comment by AngryGami** [comment]
Worked last time I've checked (2 months or so ago). Maybe something else had change since then.

**Comment by sjllls** [comment]
> Worked last time I've checked (2 months or so ago). Maybe something else had changed since then. Oh... sorry. Haven't noticed AOSP 11. So, I have no idea if IMS work in 11...

Okk, it is broken for me, I am not sure if it is related with my carrier....
Thanks for your info!
